primitives/arithmetic/src/per_things.rs

```diff
@@ -64,12 +112,43 @@ pub trait PerThing: Sized + Saturating + Copy {
 	/// `M` as the data type that can hold the maximum value of this per-thing (e.g. u32 for
 	/// perbill), this can only work if `N == M` or `N: From<M> + TryInto<M>`.
 	fn from_rational_approximation<N>(p: N, q: N) -> Self
-		where N: Clone + Ord + From<Self::Inner> + TryInto<Self::Inner> + ops::Div<N, Output=N>;
+	where N: Clone + Ord + From<Self::Inner> + TryInto<Self::Inner> + ops::Div<N, Output=N>;
+}
+
+/// Helper function to calculate a remainder used in `PerThing` multiplications.
+fn rem_helper<N, Inner, Upper>(x: N, part: Inner, acc: Inner, reciprocal: bool, truncate: bool) -> Upper 
+where 
+	N: Clone + From<Inner> + UniqueSaturatedInto<Inner> + ops::Div<N, Output=N> + ops::Mul<N, Output=N> + ops::Add<N, Output=N> + ops::Rem<N, Output=N>,
+	Inner: BaseArithmetic + Copy,
+	Upper: BaseArithmetic + Copy + From<Inner> + TryInto<Inner>,
+{
+	let (numer_upper, denom_n, denom_upper): (Upper, N, Upper) = if reciprocal {
+		(acc.into(), part.into(), part.into())
+	} else {
+		(part.into(), acc.into(), acc.into())
+	};
+	let rem = x.clone().rem(denom_n);
+	// `rem_sized` is inferior to $max, thus it fits into $type. This is assured by
```

Comment:
```suggestion
	// `rem_sized` is inferior to $max, thus it fits into `Inner`. This is assured by
```

Reply:
These documents are now a bit outdated. I am not sure what's the best way to update them though.  Maybe replace all and then name that `Inner` is always `$type` in macro expansion and `Upper` is `$upper`. 

But then someone can mess with this function and call it like 

```
rem_helper::<N, u32, u16>(
	x.clone(), 
	self.deconstruct(), 
	Self::ACCURACY, 
	false, 
	truncate,
)
```

which breaks all the assumptions. 

Hence the function must accept a generic type `P: PerThing` and then use the types `P::Upper` and `P::Inner`. Then we can assume that these assumptions hold

Reply:
>These documents are now a bit outdated.

I've rewritten them, they should make sense again. Let me know if they don't.

> Hence the function must accept a generic type P: PerThing and then use the types P::Upper and P::Inner. Then we can assume that these assumptions hold

👍